helix-core/src/main/java/org/apache/helix/HelixAdmin.java

```diff
@@ -367,6 +368,41 @@ void manuallyEnableMaintenanceMode(String clusterName, boolean enabled, String r
    */
   boolean isInMaintenanceMode(String clusterName);
 
+  /**
+   * Puts a cluster into pause mode, which will pause controller and participants.
+   * This can be used to retain the cluster state. When this method returns, it means
+   * the pause signal has been successfully sent, but it does not mean the cluster has
+   * fully entered pause mode. Because the cluster can take some time to finish the pause and
+   * process possible pending state transitions if necessary.
+   * <p>
+   * To check the cluster pause status, call {@link #getClusterPauseStatus(String)}.
+   *
+   * @param clusterName Participant cluster name.
+   * @param cancelPendingST whether or not cancel the pending state transitions for participants.
+   * @param reason The reason to put the cluster into pause mode.
+   */
+  void enableClusterPauseMode(String clusterName, boolean cancelPendingST, String reason);
+
+  /**
+   * Gets cluster pause status {@link ClusterPauseStatus}.
+   *
+   * @param clusterName cluster name.
+   * @return {@link ClusterPauseStatus}
+   */
+  ClusterPauseStatus getClusterPauseStatus(String clusterName);
```

Comment:
I suggest to make some meaningful refactoring instead of introducing more patches.
1. ClusterPauseStatus -> ClusterManagementStatus {NORMAL, CONTROLLER_PAUSED,  CLUSTER_PAUSING, CLUSTER_PAUSED, CLUSTER_UNPAUSING, CONTROLLER_MAINTENANCE}
2. No more enable or disable, just call it set status.
3. Deprecate all the other existing method to pause or maintain the cluster.

Reply:
I total agree with this direction, which I have thought about, as I designed the REST endpoint `PUT /cluster/status`.
However, I think this is out of scope of this PR as it'd need quite many changes including the code in REST. It'd be cleaner to refactor the APIs in another PR.

Reply:
Not true. If you make this change, then we are diverged from what we want to implement. It would be impossible to remove this public API once we have it in a short period.

This is a new API, we don't add APIs that known to be refactored soon.